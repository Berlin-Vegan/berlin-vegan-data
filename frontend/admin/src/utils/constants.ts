export const GASTRO: string = 'GASTRO';
export const SHOPPING: string = 'SHOPPING';

export type LocationType = typeof GASTRO | typeof SHOPPING;

const API = '/api/';
export const V1 = `${API}v1/`;
export const V2 = `${API}v2/`;
