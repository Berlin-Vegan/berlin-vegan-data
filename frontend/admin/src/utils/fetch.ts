import { getCSRFToken } from './cookie';
import { UserDispatch, TYPE_USER_LOGOUT } from '../providers/UserProvider';

export const authorizedFetch = async (
  userDispatch: UserDispatch,
  url: string,
  method: string = 'GET',
  body: object | Array<unknown> | null = null
): Promise<Response> => {
  const response = await fetch(url, {
    method,
    body: body ? JSON.stringify(body) : body,
    headers: {
      'Content-Type': 'application/json',
      'X-CSRFToken': getCSRFToken(),
    },
  });

  if (response.status === 403) {
    userDispatch({ type: TYPE_USER_LOGOUT });
  }

  return response;
};

export const fetchGastroList = async (
  userDispatch: UserDispatch,
  filter: string
): Promise<Response> =>
  authorizedFetch(userDispatch, `/api/v1/gastros/?${filter}`);
