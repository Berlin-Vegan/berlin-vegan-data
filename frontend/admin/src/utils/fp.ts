import {
  addIndex,
  compose,
  curry,
  defaultTo,
  join,
  map,
  nth,
  pipe,
  splitAt,
  toLower,
  toUpper,
} from 'ramda';

export const nthOr = curry((defaultVal, list) =>
  compose(defaultTo(defaultVal), nth)(list),
);

export const nthOrEmpty = nthOr('');

export const mapIndexed = addIndex(map);

export const capitalizeWord = pipe(
  toLower,
  splitAt(1),
  (splitted) => [toUpper(nthOrEmpty(0)(splitted)), nthOrEmpty(1)(splitted)],
  join(''),
);
