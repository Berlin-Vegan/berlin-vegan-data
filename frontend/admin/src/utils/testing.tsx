import renderer from 'react-test-renderer';
import React, { ComponentType } from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';

export const standardComponentTest = (C: ComponentType<any>, props = {}) => {
  renderWithTestRenderer(C, props);
  it('renders without crashing', () => renders(C, props));
};

export const standardComponentTestMockFetch = (
  C: ComponentType<any>,
  props = {},
  fakeResponse: any,
) => {
  renderWithTestRenderer(C, props);
  it('renders without crashing and mock fetch', async () => {
    jest.spyOn(window, 'fetch').mockImplementation((): Promise<any> => {
      const fetchResponse = {
        json: () => Promise.resolve(fakeResponse),
      };
      return Promise.resolve(fetchResponse);
    });
    await act(async () => renders(C, props));
  });
};

export const renders = (C: ComponentType, props: object) => {
  const div = document.createElement('div');
  ReactDOM.render(<C {...props} />, div);
};

const getName = (C: ComponentType) => C.displayName || C.name || 'A Component';

const renderWithTestRenderer = (C: ComponentType, props: object) => {
  it(`render <${getName(C)} />`, () => {
    const tree = renderer.create(<C {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
};
