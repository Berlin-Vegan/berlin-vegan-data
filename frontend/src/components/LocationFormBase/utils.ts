import { join, pipe, split } from 'ramda';

export const buildLabel = (key: string): string =>
  pipe(split(/(?=[A-Z])/), join(' '))(key);
