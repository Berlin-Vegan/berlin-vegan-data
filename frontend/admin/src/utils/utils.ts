import { pipe, toLower } from 'ramda';
import { LocationType, V1, V2 } from './constants';

//
export const buildReviewListUrl: string = `${V1}review/`;
export const buildReviewDetailUrl = (id: number): string =>
  `${buildReviewListUrl}${id}/`;

// API V2 urls
export const buildListUrl = (type: LocationType): string =>
  pipe(toLower, (lowerType: string) => `${V2}${lowerType}/`)(type);
export const buildDetailUrl = (type: LocationType, id: string): string =>
  pipe(toLower, buildListUrl, (baseUrl: string) => `${baseUrl}${id}/`)(type);
// FE urls
export const buildFEDetailUrl = (type: LocationType, id: string): string =>
  pipe(toLower, (lowerType: string) => `/${lowerType}/${id}/edit`)(type);
