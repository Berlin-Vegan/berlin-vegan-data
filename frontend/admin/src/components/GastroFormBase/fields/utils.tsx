import React from 'react';
import MenuItem from '@material-ui/core/MenuItem';
import { map } from 'ramda';
import { nthOr } from '../../../utils/fp';

const menuItem = (value: number | string, label: string) => (
  <MenuItem value={value} key={value}>
    {label}
  </MenuItem>
);

export const menuItems = (fieldOptions: Array<[number | string, string]>) =>
  map((item) => menuItem(nthOr('', 0)(item), nthOr('', 1)(item)))(fieldOptions);

export const withLeadingZero = (value: number): string =>
  value < 10 ? `0${value}` : `${value}`;
