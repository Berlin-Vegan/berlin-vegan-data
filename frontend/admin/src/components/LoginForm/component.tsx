import React, { Dispatch, SetStateAction, useContext, useState } from 'react';
import Avatar from '@material-ui/core/Avatar';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import { Typography } from '@material-ui/core';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import clsx from 'clsx';
import { __, propOr } from 'ramda';
import useStyles from './styles';
import {
  AuthContext,
  TYPE_SET_USER_DATA,
  TYPE_USER_LOGIN,
  UserDispatch,
} from '../../providers/UserProvider';
import { authorizedFetch } from '../../utils/fetch';

const getUserData = (userDispatch: UserDispatch) =>
  authorizedFetch(userDispatch, '/api/v1/accounts/profile/').then((res) =>
    res.json()
  );

const login = async (
  username: string,
  password: string,
  setErrors: Dispatch<SetStateAction<IErrorsState>>,
  userDispatch: UserDispatch
): Promise<void> => {
  const response = await fetch('/api/v1/accounts/login/', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ login: username, password }),
  });

  if (response.status === 200) {
    userDispatch({ type: TYPE_USER_LOGIN });
    const userData = await getUserData(userDispatch);
    userDispatch({ type: TYPE_SET_USER_DATA, payload: userData });
  } else {
    const errors = await response.json();
    const propEmpty = propOr('', __, errors);

    setErrors({
      detail: propEmpty('detail'),
      login: propEmpty('login'),
      password: propEmpty('password'),
    });
  }
};

interface IErrorsState {
  detail: string;
  login: string;
  password: string;
}

const LoginForm = () => {
  const classes = useStyles();
  const { dispatch: userDispatch } = useContext(AuthContext);
  const [formState, setFormState] = useState({ username: '', password: '' });
  const [errorsState, setErrors] = useState<IErrorsState>({
    detail: '',
    login: '',
    password: '',
  });

  const handleFormChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setFormState({ ...formState, [e.target.name]: e.target.value });
  const handleButtonClick = (e: React.MouseEvent) => {
    e.preventDefault();
    login(formState.username, formState.password, setErrors, userDispatch);
  };

  return (
    <div className={classes.paper}>
      <Avatar className={classes.avatar}>
        <LockOutlinedIcon />
      </Avatar>
      <Typography component="h1" variant="h5">
        Sign in
      </Typography>
      <form className={classes.form}>
        <Typography
          variant="subtitle1"
          className={clsx(
            !errorsState.detail && classes.noError,
            errorsState.detail && classes.error
          )}
        >
          {errorsState.detail}
        </Typography>
        <TextField
          error={!!errorsState.login}
          variant="outlined"
          margin="normal"
          required
          fullWidth
          id="username"
          label="Username"
          name="username"
          autoComplete="username"
          autoFocus
          onChange={handleFormChange}
          helperText={errorsState.login}
        />
        <TextField
          error={!!errorsState.password}
          variant="outlined"
          margin="normal"
          required
          fullWidth
          name="password"
          label="Password"
          type="password"
          id="password"
          autoComplete="current-password"
          onChange={handleFormChange}
          helperText={errorsState.password}
        />
        <Button
          type="submit"
          fullWidth
          variant="contained"
          color="primary"
          className={classes.submit}
          onClick={handleButtonClick}
        >
          Sign In
        </Button>
      </form>
    </div>
  );
};

export default LoginForm;
