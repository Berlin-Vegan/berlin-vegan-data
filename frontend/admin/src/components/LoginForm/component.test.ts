import { standardComponentTest } from '../../utils/testing';
import LoginForm from './component';

standardComponentTest(LoginForm);
