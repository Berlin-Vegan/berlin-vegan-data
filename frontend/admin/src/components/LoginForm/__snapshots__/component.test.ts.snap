// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render <LoginForm /> 1`] = `
<div
  className="makeStyles-paper-1"
>
  <div
    className="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-2 MuiAvatar-colorDefault"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
      />
    </svg>
  </div>
  <h1
    className="MuiTypography-root MuiTypography-h5"
  >
    Sign in
  </h1>
  <form
    className="makeStyles-form-3"
  >
    <h6
      className="MuiTypography-root makeStyles-noError-5 MuiTypography-subtitle1"
    >
      
    </h6>
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink={false}
        htmlFor="username"
        id="username-label"
      >
        Username
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="username"
          autoFocus={true}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="username"
          name="username"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-9"
          >
            <span>
              Username
               *
            </span>
          </legend>
        </fieldset>
      </div>
      
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink={false}
        htmlFor="password"
        id="password-label"
      >
        Password
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="current-password"
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="password"
          name="password"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="password"
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-9"
          >
            <span>
              Password
               *
            </span>
          </legend>
        </fieldset>
      </div>
      
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-4 MuiButton-containedPrimary MuiButton-fullWidth"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Sign In
      </span>
    </button>
  </form>
</div>
`;
